File: steps/5/main/scala/eightyDays/scala/account/fee/LowBalancePerBooking.scala
package eightyDays.scala.account.fee

import java.time.LocalDateTime
import eightyDays.scala.account.{Account, Amount}

trait LowBalancePerBooking extends Account {
  val fee: Amount

  val threshold: Amount

  private[this] def feeIfPoossible(account: Account, valuta: LocalDateTime) =
    if (account.balance < threshold)
      account.post(-fee, valuta)
    else
      account

  override def withdraw(value: Amount, valuta: LocalDateTime): Account = feeIfPoossible(super.withdraw(value, valuta), valuta)

  override def deposit(value: Amount, valuta: LocalDateTime): Account = feeIfPoossible(super.deposit(value, valuta), valuta)
}

File: steps/7/main/java/eightyDays/java/account/fee/LowBalancePerBooking.java
package eightyDays.java.account.fee;

import eightDays.java.account.Account;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.function.BiFunction;

public class LowBalancePerBooking {
    private final BigDecimal fee;
    private final BigDecimal threshold;
    private final BiFunction<BigDecimal, LocalDateTime, Account> predecessor;

    public LowBalancePerBooking(BiFunction<BigDecimal, LocalDateTime, Account> pPredecessor, BigDecimal pThreshold, BigDecimal pFee) {
        threshold = pThreshold;
        fee = pFee;
        predecessor = pPredecessor;
    }

    public BiFunction<BigDecimal, LocalDateTime, Account> build() {
        return (value, valuta) -> {
            Account booked = predecessor.apply(value, valuta);
            if (booked.getBalance().compareTo(threshold) <= 0)
                return booked.post(fee.negate(), valuta);
            else
                return booked;
        };
    }
}
